pkg/block/block.go

```diff
@@ -156,6 +156,11 @@ func deleteDir(ctx context.Context, logger log.Logger, bkt objstore.Bucket, dir
 		if strings.HasSuffix(name, objstore.DirDelim) {
 			return deleteDir(ctx, logger, bkt, name)
 		}
+		metaFile := path.Join(dir, MetaFilename)
```

Comment:
OK while I see the problem, but do you this exact fix here really makes sense?

`deleteDir` is responsible for deleting a "directory" in object storage, right? So this is what the user expects. Now we suddenly ignore meta.json file some reason. Do you think this is safe? (: I think this will be extremely surprising and might introduce bug someday when someone will try to reuse this function for something else than now... And we love to avoid surprises in Thanos project! (: WDYT? You should never assume usage without explicitly mentioning it in name/comment. ):

In this case, what would work is to pass to `deleteDir` some `keep func(name string)bool` argument, so users can decide on their own. Also can we rename `deleteDir` to `deleteDirRec` to be more explicit :hugs:

Reply:
Having a whole function to decide on what file might stay or not, might overcomplicate things though. What about a slice of strings? 
Also, what crossed my mind: what about ignoring the 404 in swift? The contract of `objstore.swift.Delete` is that it should remove the object, the 404 tells you it's not there, so the contract is fulfilled.

Reply:
Why overcomplicate? It will have fewer lines then what you added now (: 

```
if keep(name) {
   return nil 
}
```

Reply:
> what about ignoring the 404 in swift?

Ignornig on what level?

Reply:
We like 404 to be an error, as you remove something which you expect to be there. We can consider and discuss it, but for here it will super easy to add `keep` function (: